### DISCUSSION TITLE
Getting Discord Guilds

### DISCUSSION THREAD

**Discussion by nixxxo** [nextauthjs/next-auth#3774] [Help]
### Question 💬

Hello,
I've been reading the docs and trying stuff out for sometime now, with callbacks, changing scopes, playing with prisma and so one. Despite all of this, I just can't figure out how to log the guilds the user is in. I tried using the access token to send like a "second" fetch request to "https://discordapp.com/api/users/@me/guilds" instead of the default "https://discordapp.com/api/users/@me", however it doesn't really work to fetch information withing the NextAuth().

So my question is how can I log every server the user is in, the moment he authenticates with discord?

p.s I know UseState() cannot be there but honestly just figuring out where it has to be might fix the issue for me, just how to write the guilds to a variable and from there on i can write them to the db and do whatever i want.

Any help is appreciated.

### How to reproduce ☕️

```js
import NextAuth from "next-auth"
import DiscordProvider from "next-auth/providers/discord"
import CredentialsProvider from "next-auth/providers/credentials";
import { PrismaAdapter } from "@next-auth/prisma-adapter"
import { PrismaClient } from "@prisma/client"
import { useState } from "react";
const prisma = new PrismaClient()



export default NextAuth({
  adapter: PrismaAdapter(prisma),
  callbacks: {
    async signIn({ user, account, profile, email, credentials }) {
      const access_token = account.access_token
      const [data,setData]=useState([]);
      fetch("https://discordapp.com/api/users/@me/guilds", {
        headers: {
          Authorization: "Bearer "+access_token,
          "Content-Type": "application/json"
        }
      })
        .then(function(response){
          console.log(response)
          return response.json();
        })
        .then(function(myJson) {
          console.log(myJson);
          setData(myJson);
        });
      console.log(guilds)
      return true
    },
  },
  providers: [
    DiscordProvider({
      clientId: process.env.DISCORD_CLIENT_ID,
      clientSecret: process.env.DISCORD_CLIENT_SECRET,
      authorization: { params: { scope: 'identify guilds email' } },
      profile(profile) {
        if (profile.avatar === null) {
          const defaultAvatarNumber = parseInt(profile.discriminator) % 5
          profile.image_url = `https://cdn.discordapp.com/embed/avatars/${defaultAvatarNumber}.png`
        } else {
          const format = profile.avatar.startsWith("a_") ? "gif" : "png"
          profile.image_url = `https://cdn.discordapp.com/avatars/${profile.id}/${profile.avatar}.${format}`
        }
        return {
          id: profile.id,
          name: profile.username,
          email: profile.email,
          image: profile.image_url,
        }
      },
    })
    // ...add more providers here
  ],
  database: process.env.DATABASE_URL,
  secret:process.env.SECRET_KEY_OPENSSL
})


```

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by ryanbarr** [comment]
Hey @nixxxo,

Before I try to provide some guidance, I'd like to make sure I understand what you're trying to accomplish:

1. When a user logs in using the `Discord` provider, then
2. Immediately make a request to the Discord API for the user's Guilds, then
3. Store those guilds in the database.

Assuming that the above requirements are what you're looking for, and also assuming that the `signIn()` callback you've provided is functioning as desired:

1. In your `signIn()` callback, make an `await fetch()` GET to the Discord API to retrieve the user's guilds, then
2. Handle that response accordingly and make an `await fetch()` POST/PUT/PATCH to your API to store the guilds in relation to the user, then
3. `return true;` to allow the user to complete the sign-in.

Following this process will ensure that both the request to the Discord API and the save to your API complete in full before allowing the user to sign-in. Additionally, you can decide to `return false;` in the event that either of the calls did not succeed or, for example, you want to ensure that the user is a member of a specific guild before authorizing sign-in.

Hope that helps!

**Comment by TristonWasik** [comment]
After doing some digging, I found that you can just overwrite the authorization property in the DiscordProvider method:

```
providers: [
    DiscordProvider({
      clientId: env.DISCORD_CLIENT_ID,
      clientSecret: env.DISCORD_CLIENT_SECRET,
      authorization:
        "https://discord.com/api/oauth2/authorize?scope=identify+email+guilds",
    }),
  ],
```

After doing the above, I was able to access the users guilds using the "/api/users/@me/guilds" endpoint with no problems.

**Reply by danielterwiel** [reply]
@TristonWasik can I ask how got the guilds data to the client? I got the list inside the signIn callback, but I see no other option but to temporary store them in the database, which is suboptimal
